src/cursor/cur_version.c

```diff
@@ -0,0 +1,685 @@
+
+    if (!F_ISSET(version_cursor, WT_CURVERSION_UPDATE_EXHAUSTED)) {
+        /*
+         * If we position on a key, set next update of the version cursor to be the first update on
+         * the key if any.
+         */
+        page = cbt->ref->page;
+        switch (page->type) {
+        case WT_PAGE_ROW_LEAF:
+            if (cbt->ins != NULL)
+                upd = cbt->ins->upd;
+            else {
+                rip = &page->pg_row[cbt->slot];
+                upd = WT_ROW_UPDATE(page, rip);
+            }
+            break;
+        case WT_PAGE_COL_FIX:
+        case WT_PAGE_COL_VAR:
+            if (cbt->ins != NULL)
+                upd = cbt->ins->upd;
+            else
+                upd = NULL;
+            break;
+        default:
+            WT_ERR(__wt_illegal_value(session, page->type));
+        }
+
+        /* Skip aborted updates. */
+        while (
+          version_cursor->next_upd != NULL && version_cursor->next_upd->txnid == WT_TXN_ABORTED)
+            version_cursor->next_upd = version_cursor->next_upd->next;
+
+        /* Nothing to check. */
+        if (version_cursor->next_upd == NULL)
+            upd = NULL;
+
+        /*
+         * Walk to the next upd from start each time to ensure the next upd is not obsolete or
+         * freed. Check aborted after checking next upd because the next upd can be aborted
+         * concurrently. In this case, we will return an extra aborted update, which should be fine.
+         */
+        for (; upd != NULL; upd = upd->next) {
+            /* We walk to the desired update. */
+            if (upd == version_cursor->next_upd)
+                break;
+
+            if (upd->txnid == WT_TXN_ABORTED)
+                continue;
+
+            /* We have traversed all the non-obsolete updates. */
+            if (WT_UPDATE_DATA_VALUE(upd) && __wt_txn_upd_visible_all(session, upd)) {
+                upd = NULL;
+                break;
+            }
+        }
+
+        if (upd == NULL) {
+            version_cursor->next_upd = NULL;
+            F_SET(version_cursor, WT_CURVERSION_UPDATE_EXHAUSTED);
+        } else {
+            if (upd->type == WT_UPDATE_TOMBSTONE) {
+                tombstone = upd;
+
+                /*
+                 * If the update is a tombstone, we still want to record the stop information but we
+                 * also need traverse to the next update to get the full value. If the tombstone was
+                 * the last update in the update list, retrieve the ondisk value.
+                 */
+                version_cursor->upd_stop_txnid = upd->txnid;
+                version_cursor->upd_durable_stop_ts = upd->durable_ts;
+                version_cursor->upd_stop_ts = upd->start_ts;
+
+                /* No need to check the next update if the tombstone is globally visible. */
+                if (__wt_txn_upd_visible_all(session, upd))
+                    upd = NULL;
+                else
+                    upd = upd->next;
```

Comment:
This kind of bothers me -- if a version cursor is used to debug the list of updates, but we trim the list of updates because of our processing rules, doesn't that mess up the whole idea?

Reply:
I think that should be fine. We are just ignoring updates that are obsolete. A reader should also not reach these updates.

Reply:
Yes, but "debug" should list everything? Anyway, your choice!

Reply:
Agree, ideally we should list everything but obviously we cannot achieve that now.